### ISSUE TITLE
autotune: Anti-gaming composite evaluation metric

### ISSUE THREAD

**Issue by akaszubski** [akaszubski/ha-victron-mpc#46]
## Summary

Design and implement the composite scalar metric that scores parameter configurations during autotune. This is the hardest and most important piece — the metric must accurately reflect real-world cost while being impossible for the optimization loop to game.

## The Core Problem

The autotune loop (see #44) modifies LP tunables to minimize a cost metric. If the metric uses the tunable values themselves for scoring, the optimizer can game it:

| Gaming Vector | Example | Result |
|--------------|---------|--------|
| Zero wear cost | Set `battery_wear_cost=0` → LP cycles aggressively → metric shows lower cost | Battery degrades, real cost higher |
| Remove floor penalty | Set `soft_floor_penalty=0` → LP ignores 30% floor → metric shows savings | Battery drained to unsafe levels |
| Inflate SoC profile | Set `soc_profile_pre_peak=0.40` → LP always grid-charges → metric counts "high-value" battery | Over-charges at grid rates above actual benefit |

**Solution**: The evaluation metric uses FIXED penalty costs that are independent of the tunables being optimised. The tunables affect LP *decisions*, but the *scoring* uses immutable constants.

## Architecture

```
train_config.json (tunables) → optimizer.py (LP decisions) → metric.py (FIXED scoring)
       ↑ agent edits                                              ↑ agent CANNOT edit
```

The tunables influence *what the LP does* (charge/discharge/hold timing). The metric scores *what actually happened* using fixed, real-world costs.

### `autotune/metric.py`
Composite metric calculation with fixed evaluation constants:

```python
# FIXED evaluation costs — NOT from tunables, cannot be gamed
EVAL_WEAR_COST = 0.02                    # $/kWh — actual Pylontech LFP wear
EVAL_FLOOR_VIOLATION_PENALTY = 0.50      # $/violation-step (per 5-min step below 30%)
EVAL_SUNSET_PENALTY = 0.10              # $/% below 80% at sunset
EVAL_SOC_CONTINUITY_PENALTY = 0.05      # $/% deviation start→end over full period
EVAL_EXCESSIVE_CYCLING_PENALTY = 0.01   # $/kWh above 1 full cycle/day
```

**Composite metric formula** (total AUD, lower is better):

```
metric = Σ over all days:
  + grid_import_kwh × actual_buy_price              (real grid cost)
  - grid_export_kwh × actual_sell_price              (real export revenue)
  + discharge_kwh × EVAL_WEAR_COST                   (FIXED wear, not tunable)
  + floor_violation_steps × EVAL_FLOOR_VIOLATION_PENALTY
  + max(0, 80 - sunset_soc_pct) × EVAL_SUNSET_PENALTY   (per day with sunset)

+ over full period:
  + |end_soc% - start_soc%| × EVAL_SOC_CONTINUITY_PENALTY
  + max(0, total_discharge_kwh - days × capacity) × EVAL_EXCESSIVE_CYCLING_PENALTY
```

Functions:
- `compute_day_metrics(opt_output: OptOutput, day_data: DayData, start_soc_kwh: float) -> DayResult`
- `compute_composite_metric(day_results: list[DayResult], start_soc_kwh: float, end_soc_kwh: float) -> EvalResult`

### `autotune/metric_config.py`
Fixed evaluation constants in a separate file for clarity. Listed in `program.md` as immutable.

### `autotune/evaluate.py` (from #44)
Wire in the composite metric from `metric.py` instead of raw grid cost. The evaluate CLI should output the full `EvalResult` as JSON.

## Anti-Gaming Properties

| Property | How enforced |
|----------|------------|
| Wear cost fixed at $0.02 | EVAL_WEAR_COST is in metric.py (agent can't touch), not from tunables |
| Floor violations per-step | Counted per 5-min step (288/day), not per day — can't game with 1 bad day |
| Sunset target fixed at 80% | Uses 80% threshold regardless of tunable `sunset_soc_target_pct` |
| SoC continuity | Start-to-end SoC delta over full 14-30 day period — can't drain on last day |
| Excessive cycling | Penalty above 1 full cycle/day prevents "always cycle" strategies |
| Multi-day window | 14-30 days captures weather/price diversity — can't overfit to one pattern |

## Testing

Create `tests/test_autotune_metric.py`:

### Anti-gaming tests (critical):
- [ ] Setting `battery_wear_cost=0` in tunables does NOT eliminate wear from metric (EVAL_WEAR_COST still applies)
- [ ] Setting `soft_floor_penalty=0` in tunables does NOT remove floor violation penalty from metric
- [ ] Setting extreme SoC profile values doesn't reduce metric below physical cost floor

### Correctness tests:
- [ ] Floor violation counting: SoC at 25% for 12 steps = 12 × $0.50 = $6.00 penalty
- [ ] Sunset penalty: SoC at 60% at sunset = (80-60) × $0.10 = $2.00 penalty
- [ ] SoC continuity: start 50%, end 30% over 14 days = 20 × $0.05 = $1.00 penalty
- [ ] Excessive cycling: 2 full cycles/day for 14 days = 14 × 14.2kWh × $0.01 = $1.99 penalty
- [ ] Determinism: same inputs always produce same metric (no randomness)

### Sensitivity tests:
- [ ] Metric changes by >1% when key parameters change by 20%
- [ ] Metric is monotonic in grid cost (more grid import = higher metric)
- [ ] Metric is monotonic in wear (more cycling = higher metric)

## Acceptance Criteria

- [ ] Metric is deterministic: same tunables + same data = same scalar output, every time
- [ ] Setting `battery_wear_cost=0.0` in tunables changes LP behavior but does NOT reduce evaluation wear penalty
- [ ] Floor violations incur $0.50 per 5-min step (measurable, significant)
- [ ] Multi-day evaluation (14+ days) prevents single-day boundary gaming
- [ ] Metric is a single float (total AUD cost) — lower is better
- [ ] All anti-gaming tests pass
- [ ] Sensitivity tests confirm metric responds to meaningful parameter changes

## Dependencies

- #44 (Phase 1) — evaluate.py, types.py (DayResult, EvalResult)
- #45 (Phase 2) — cached test data for realistic metric validation

## Risks

| Risk | Mitigation |
|------|------------|
| Overfitting to historical data | Require evaluation dataset spans ≥14 days with weather diversity. Future: train/validation split. |
| Fixed wear cost diverges from reality | EVAL_WEAR_COST is conservative ($0.02); actual Pylontech LFP wear may be lower. Review annually. |
| Metric insensitive to some parameters | Sensitivity tests in this phase identify low-impact parameters. Remove them from train_config.json. |
| Sunset penalty during cloudy days | Only apply sunset penalty on days where solar_forecast > 5kWh (enough sun to charge). |

## Effort

~1 sprint (2 new files + 1 modified + 1 test file)
